### ISSUE TITLE
Zoning in a settlement

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#720]
**Describe the bug**
- We have not modeled the concept of zoning in a settlement and how to define different priority of traffic between them.
- Currently, from time to time, settlers would go to the dining hall in Astronomy Observatory to eat, to do research, or to sleep. This create higher and unneeded traffic via EVA airlocks.

**Expected and unexpected behavior**
- In case of *Alpha Base* settlement template, a majority of the settlement is in one zone, while the astronomy observatory and its attached EVA airlock sit in another physically isolated zone. 

**Discuss any potential problems you have and have not ruled out**
- One temporary mitigation to the problem of zoning is the use of an underground tunnel between the main zone and the astronomy zone. 
- The tunnel building exits but the surface collision still applies to tunnel. So currently, it's not considered an underground structure yet.

**Screenshots**
![image](https://user-images.githubusercontent.com/1168584/194734309-7e94388d-4304-421d-a27c-72908490a08a.png)

![image](https://user-images.githubusercontent.com/1168584/194734330-f7922099-b365-4b9f-b754-3717068785e9.png)

**Version**
 - pre-3.4.0 build 7474

**Additional context**

**Comment by mokun** [comment]
@bevans2000 

One temporary fix is to let settlers grab some dessert or dry food and carry them whenever they go.  When they are hungry, they could eat.

**Comment by mokun** [comment]
Shall we give each building a zone id ? 

A zone id will help a lot when we search for a particular building function to walk to. It will search and select for buildings within the same zone. If not found, it will search the adjacent zone.

**Comment by bevans2000** [comment]
Yes that could work so when looking for a place to eat we only go to the same zone. But we would only using zoning for certain task, ie those general ones. They are many cases where changing zone is essential, eg in the Observatory now but going to Tend Crops, I need to find a farm in a different zone.

**Comment by mokun** [comment]
It's tempting to just connect the observatory with the main zone with tunnel and be done with it.

But getting the idea of zoning implemented will be good for the long run.

So I'll need to start with those settlement templates.

**Comment by bevans2000** [comment]
Fine with the Template, how do you intend to use them in the Tasks? Some Tasks will need to look for Building within the same Zone and some can go to any Zones. Need to be careful we have a flexible approach.

**Comment by mokun** [comment]
Btw, just saw a person leaving an airlock in order to go to Astronomy Observatory just to do the Reading task.

This is NOT supposed to happen. 

When the reading task is done, this person left the the Observatory and had to go through all the trouble of walking outside back to the main zone of the settlement.

It seems that in Read task, the Building instance returned by calling getAvailableDiningBuilding() or getAvailableRecreationBuilding() can be Observatory. 

The zone filtering done earlier fail to filter Observatory from the list.

**Comment by bevans2000** [comment]
The concept of zoning should work so it must be a bug somewhere

**Comment by bevans2000** [comment]
Yes sure
